sympy/angles/euler.py

```diff
@@ -0,0 +1,688 @@
+
+        Examples
+        ========
+
+        >>> from sympy import Euler
+        >>> from sympy import pi
+        >>> angles = Euler(pi/2, 0, 0, seq='ZYX')
+        >>> angles
+        Euler(pi/2,  0,  0, ZYX)
+
+        >>> angles.to_rotation_matrix()
+        Matrix([
+        [0, -1, 0],
+        [1,  0, 0],
+        [0,  0, 1]])
+
+        Generates a 4x4 transformation matrix (used for rotation about a point
+        other than the origin) if the point(v) is passed as an argument.
+
+        Examples
+        ========
+
+        >>> from sympy import Euler
+        >>> from sympy import pi
+        >>> angles = Euler(pi/2, 0, 0, seq='ZYX')
+        >>> angles.to_rotation_matrix((1, 1, 1))
+        Matrix([
+        [0, -1, 0, 2],
+        [1,  0, 0, 0],
+        [0,  0, 1, 0],
+        [0,  0, 0, 1]])
+
+        """
+
+        funcs = [rot_axis1, rot_axis2, rot_axis3]
+        # negative because the functions are defined in left-hand convention
+        Ri = funcs[self.i - 1](- self.alpha)
+        Rj = funcs[self.j - 1](- self.beta)
+        Rk = funcs[self.k - 1](- self.gamma)
+
+        if self.extrinsic:
+            M = Rk * Rj * Ri
+        else:
+            M = Ri * Rj * Rk
+
+        if not v:
+            return M
+
+        else:
+            m00, m01, m02, m10, m11, m12, m20, m21, m22 = M.flat()
+            (x, y, z) = v
+
+            m03 = x - x*m00 - y*m01 - z*m02
+            m13 = y - x*m10 - y*m11 - z*m12
+            m23 = z - x*m20 - y*m21 - z*m22
+            m30 = m31 = m32 = 0
+            m33 = 1
+
+            return Matrix([[m00, m01, m02, m03], [m10, m11, m12, m13],
+                          [m20, m21, m22, m23], [m30, m31, m32, m33]])
+
+    @classmethod
+    def from_rotation_matrix(cls, M, seq):
+        """Returns the equivalent Euler angles of a matrix.
+
+        Parameters
+        ==========
+
+        M : Matrix
+            Input matrix to be converted to equivalent quaternion.
+
+
+        seq : string of length 3
+            Represents the sequence of rotations.
+            For intrinsic rotations, seq must be all lowercase and its elements
+            must be from the set `{'x', 'y', 'z'}`
+            For extrinsic rotations, seq must be all uppercase and its elements
+            must be from the set `{'X', 'Y', 'Z'}`
+
+        Returns
+        =======
+
+        Euler
+            The Euler angles equivalent to given matrix.
+
+        Examples
+        ========
+
+        >>> from sympy import Euler
+        >>> from sympy import Matrix, symbols
+        >>> x = symbols('x')
+        >>> M = Matrix([[0, -1, 0], [1, 0, 0], [0, 0, 1]])
+        >>> euler = Euler.from_rotation_matrix(M, 'ZYX')
+        >>> euler
+        Euler(pi/2,  0,  0, ZYX)
+
+        """
+        q = Quaternion.from_rotation_matrix(M)
+        q = trigsimp(q)
+        return Euler.from_quaternion(q, seq)
+
+    @staticmethod
+    def _generic_mul(lhs, rhs):
+        """Generic multiplication.
```

Comment:
If adding doesn't really make sense then what does multiplication mean?

Reply:
I thought of implementing multiplication as a composition of rotations, just like for a rotation matrix and for a quaternion.

Except, of course, that most of the time Euler angles are horrible for composing, this is why I implemented it as a multiplication of quaternions followed by a conversion to Euler angles.